Skeleton for NoOp restore + NuGet.exe impl

nkolev92 opened the pull request:
This PR includes:

basic skeleton for the no-op restore
NuGet.exe force implementation

Parts still missing:
MSBuild/dotnet.exe force flag
cache path building for project.json
sources in the dg spec
tests for the new feature
tool restores
& VS related No-Op changes

I will update the design doc with the latest changes in the implementation.
https://github.com/NuGet/Home/wiki/NuGet-Restore-No-Op

nkolev92 committed:
Initial impl

nkolev92 committed:
Fix namespace

nkolev92 committed:
Merge branch 'dev' into dev-nkolev92-noopcli

Commit:
Fixing typo from previous commit that was blocking build

Commit:
set vsix build number to the insertable version (#1333)

nkolev92 committed:
Fix nuget.exe restore

mishra14 committed:
Fixing broken ThreadDispatcher in Test code that was blocking unit tests (#1334)

nkolev92 committed:
add resource file

Commit:
add a script to publish artifacts from vsts to myget (#1336)

sbanni committed:
Accessibility bug fixes (#1337)

* Give default package management an accessible name

* Don't report LoadingStatusIndicator as an item in the packages list with narrator

* Set automationproperties.name for packagesource combobox

* Change accessibility name for the text box when the status changes

* Fix tab index order of the NuGet Package manager options page

mishra14 committed:
Passing CurrentDirectory to Locals Settings (#1330)

nkolev92 committed:
remove redundant file

nkolev92 committed:
fix compilation problems

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:
```diff
@@ -47,6 +47,9 @@ public class RestoreCommand : DownloadCommandBase
         [Option(typeof(NuGetCommand), "RestoreCommandRecursive")]
         public bool Recursive { get; set; }
 
+        [Option(typeof(NuGetCommand), "ForceRestoreCommand")]
```
This needs a resource string for the command line help describing this option

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs:
```diff
@@ -74,12 +74,19 @@ private IReadOnlyList<RestoreSummaryRequest> GetRequestsFromItems(RestoreArgs re
+                {
+                    projectDependencyGraphSpec.AddProject(spec.Clone()); //TODO - Is cloning the best idea here? Does it change?
+                }
+                projectDependencyGraphSpec.AddRestore(projectNameToRestore);
```
Consider adding a helper method ``DependencyGraphSpec.WithProjectClosure(string uniqueName)``, this could be used other places too such as for per project restore in VS.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:
```diff
@@ -0,0 +1,31 @@
+﻿using System;
```
license header

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    class NoOpRestoreResult : RestoreResult
+    {
+        public NoOpRestoreResult(bool success, IEnumerable<RestoreTargetGraph> restoreGraphs, IEnumerable<CompatibilityCheckResult> compatibilityCheckResults, IEnumerable<MSBuildOutputFile> msbuildFiles,
```
I would expect this to take only the needed parameters. It looks like most of these are passed in as null. It should be expected that a noop result will never have graphs or compatibility results. Anything that can be inferred from being a noop should be here.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:
```diff
@@ -0,0 +1,31 @@
+
+        //We override this method because in the case of a no op we don't need to update anything
+        public override async Task CommitAsync(ILogger log, CancellationToken token)
+        {
```
Should this log a message here?

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs:
```diff
@@ -62,6 +62,10 @@ public class RestoreRequest
             RestoreOutputPath = Path.GetDirectoryName(Project.FilePath);
         }
 
+        public DependencyGraphSpec dependencyGraphSpec { get; set; }
```
**D**ependencyGraphSpec

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -147,10 +169,40 @@ public Task<RestoreResult> ExecuteAsync()
+            {
+                cacheFile = CacheFileFormat.Read(_request.Project.RestoreMetadata.CacheFilePath, _logger);
+
+                if (cacheFile.IsValid && StringComparer.OrdinalIgnoreCase.Equals(cacheFile.DgSpecHash, newDgSpecHash))
```
Should this be Ordinal instead of OrdinalIgnoreCase? I wouldn't expect a hash to be case insensitive.

emgarten commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -511,4 +511,10 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
+    <value>The dependency graph spec for '{0}' has changed. Continuing restore.</value>
+  </data>
+  <data name="Log_RestoreNoOpFinish" xml:space="preserve">
+    <value>The dependency graph spec for '{0}'has not changed. No further actions are required to complete the restore.</value>
```
'{0}' has

emgarten replied:
``project restore inputs`` might make more sense to users than ``dependency graph spec`` which more of an internal thing.

Since this is localized I assume this intended for users. If not I would move this to the debug level and make it non-localized so that it is clear to us when users paste diagnostic output.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -0,0 +1,44 @@
+﻿using System;
```
license header

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            var combiner = new HashCodeCombiner();
+            combiner.AddObject(DgSpecHash);
+            return combiner.CombinedHash;
```
nit: return HashCodeCombiner.GetHash(DgSpecHash) would give you the same result

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -0,0 +1,44 @@
+                return false;
+            }
+
+            if (ReferenceEquals(this, other))
```
👍

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace NuGet.ProjectModel
+{
+    public class CacheFile : IEquatable<CacheFile>
```
Add a format=1 to allow versioning this file.
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L296

In IsValid verify that the version matches the expected version in code. This will help future proof this file.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    public class CacheFile : IEquatable<CacheFile>
+    {
+        public string DgSpecHash { get; set; }
```
Since this currently only holds the hash I would remove the set here and pass this into the constructor. Having this as an immutable class would simplify things.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs:
```diff
@@ -0,0 +1,120 @@
+            return Read(filePath, NullLogger.Instance);
+        }
+
+        public static CacheFile Read(string filePath, ILogger log)
```
Load might be a better name for loading from disk, I'm thinking of XDocument as an example

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs:
```diff
@@ -0,0 +1,120 @@
+            catch (Exception ex)
+            {
+                log.LogError(string.Format(CultureInfo.CurrentCulture,
+                    "error reading a cache file {1} : {2}",
```
This should be localized

emgarten replied:
Logging this as an error may be confusing since it doesn't look like this will actually fail the restore. I would suggest making this a warning or diagnostic output. Worst case we just run the restore.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs:
```diff
@@ -0,0 +1,120 @@
+            {
+                using (var jsonReader = new JsonTextReader(reader))
+                {
+                    while (jsonReader.TokenType != JsonToken.StartObject)
```
Any reason this doesn't load the JsonTextReader directly into a JObject using JObject.Parse (or one of those methods)?

emgarten replied:
Leave it if it works

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFileFormat.cs:
```diff
@@ -0,0 +1,120 @@
+            return item != null ? new JValue(item) : JValue.CreateNull();
+        }
+
+        private static JObject WriteCacheFile(CacheFile cacheFile)
```
GetCacheFile

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreResult.cs:
```diff
@@ -190,5 +190,13 @@ public ISet<LibraryRange> GetAllUnresolved()
+
+        protected async Task CommitCacheFileAsync(ILogger log)
+        {
+            log.LogDebug($"Writing cache file to disk. Path: {CacheFilePath}");
```
If the restore failed the cache file should be deleted instead of writing out the current hash. Potentially a user could add the needed packages to their source and restore would continue nooping without picking them up.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs:
```diff
@@ -50,6 +50,9 @@ public class ProjectRestoreMetadata : IEquatable<ProjectRestoreMetadata>
         public string PackagesPath { get; set; }
 
+        public string CacheFilePath {
+            get { return OutputPath + "project.nuget.json"; }  }
```
Is this going to use project.assets.json.cache? Is this a temp name still?

nkolev92 replied:
temp name...Rob suggested project.nuget.cache...the json thing is probably not useful for the user...they shouldn't need to look at this file at all.

emgarten replied:
The folder could have multiple projects, so it needs to match the project.assets.json name or include the project name.

nkolev92 replied:
I had a discussion with @rohit21agrawal, @joelverhagen and @mishra14 and they all share the opinion that it should live in the obj folder for both PJ + Pack Ref and should include the projectName. 

Based on @emgarten's feedback above, the final name will follow the:
projectName.csproj.nuget.cache pattern.

emgarten replied:
I don't believe the obj folder is read yet for project.json, you will need to add this to VS and msbuild. You may want to leave this feature off for project.json until everything else is done.

emgarten reviewed (changes requested):
Adding feedback. I realize this is still a WIP

Commit:
add pdbgit nuget package to all shiping projects (#1339)

Commit:
fix localization issue (#1340)

nkolev92 committed:
Address the feedback

nkolev92 committed:
cleanup cachefile/cache file format.

nkolev92 committed:
Address feedback

nkolev92 committed:
Change the Vsix Name to NuGet.Tools.vsix so that the packageid.json manifest file has the correct file name (#1342)

nkolev92 committed:
fix resource

nkolev92 committed:
temp test fix

Commit:
get the exact list of DLLs and PDBs that need to be indexed (#1344)

emgarten committed:
Removing unneeded caching from dependency walker

nkolev92 committed:
remove redundant code

tmds committed:
Set execute permission when extracting files.

nkolev92 committed:
feedback + cleanup

nkolev92 committed:
add validation for the targets/props and lock files

nkolev92 commented:
🔔 @emgarten  I have made a bunch more changes. There are still things missing however.
MSBuild/dotnet.exe force flag
cache path building for project.json
A lot of tests are failing. Want to discuss that offline. 
@ Others, this is still WIP, feel free to take a look & give feedback, I can ping you when everything is ready.

nkolev92 committed:
preserve old behavior

nkolev92 committed:
cleanup, extra no op logic in a separate class

nkolev92 committed:
test fix

nkolev92 committed:
cleanup extra spaces

nkolev92 committed:
fix typo

nkolev92 committed:
fix yet another typo

nkolev92 committed:
fix null check bug

nkolev92 committed:
fix bug

nkolev92 committed:
don't include the g for project.json projects

nkolev92 committed:
Don't set standard metadata if it's not been initialized yet

nkolev92 committed:
Cleanup

nkolev92 commented:
@emgarten @mishra14 @zhili1208 @rohit21agrawal @alpaix :bell:

MSBuild/dotnet.exe force flag
cache path building for project.json
sources in the dg spec
tests for the new feature 
tool restores
& VS related No-Op changes
 are not part of this PR.

I will update the design doc with the latest changes in the implementation.
https://github.com/NuGet/Home/wiki/NuGet-Restore-No-Op

emgarten committed:
RestoreTask support for ICancelableTask

ICancelableTask allows restore to be canceled from the command line using ctrl+c. This is important for large restores and scenarios where package downloads are taking a long time. Using the cancellation token this will gracefully stop in the same way Visual Studio allows canceling restore.

Fixes https://github.com/NuGet/Home/issues/4517

emgarten committed:
Avoid restore errors for projects that cannot be read by restore

Skip projects that could not be read by restore, these are commonly broken projects or projects that do not reference common targets that allow importing the restore task.

NuGet 3.5.0 and NuGet for Visual Studio ignores these projects, this adds the same behavior to msbuild.

Errors due to these invalid projects will be reported at build time. Previous attempts at warning for these scenarios in Visual Studio caused a lot of unneeded noise and had to be removed.

Fixes https://github.com/NuGet/Home/issues/4650

emgarten committed:
Throw a helpful message if HOME is not set

This change adds a helpful message when the user HOME path cannot be found. Previously CI scenarios which were not running under normal user accounts would throw exceptions around a NRE with Path.Combine when HOME was null.

Fixes https://github.com/NuGet/Home/issues/4671

emgarten committed:
Exclude project references with ReferenceOutputAssembly=false from Visual Studio restore

Projects with ReferenceOutputAssembly=false should be ignored during restore in Visual Studio. Transitive dependencies should not flow from these projects to the parent project.

Fixes https://github.com/NuGet/Home/issues/4992

Commit:
add a vsmanproj to generate a vsman file for insertion (#1348)

emgarten commented on src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:
```diff
@@ -5336,4 +5336,7 @@ nuget locals global-packages -list</value>
     <value>Restore all referenced projects for UWP and NETCore projects. This does not include packages.config projects.</value>
   </data>
+  <data name="ForceRestoreCommand" xml:space="preserve">
+    <value>NuGet restore does not perform a full restore when the inputs have not changed. This option skips the that optimization.</value>
```
This should mention that it is only for PackageReference/project.json and give some hints as to why this should be used. At the same time it needs to be clear that this isn't the same as -NoCache which is for the http cache only.

How about:
Forces all dependencies to be resolved even if the last restore was successful. This is equivalent to deleting project.assets.json. (Does not apply to packages.config)

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreResult.cs:
```diff
@@ -55,6 +56,15 @@ public class RestoreResult : IRestoreResult
+        /// <summary>
+        /// Cache File path. The file path where the cache is written out
+        /// </summary>
+        public string CacheFilePath { get;  }
```
I don't see any consumers of CacheFile or CacheFilePath. It could be useful in the future, but for now I suggest removing these public properties and keeping the cache file as an internal implementation detail. Callers should trust that restore does the right thing in terms of caching, and that the noop flag is followed.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreResult.cs:
```diff
@@ -107,62 +121,56 @@ public ISet<LibraryRange> GetAllUnresolved()
-                        log.LogDebug($"Writing tool lock file to disk. Path: {result.LockFilePath}");
+                        
+
+                        log.LogDebug(string.Format(CultureInfo.CurrentCulture,
```
debug is typically non-localized, but I don't think it is a problem to localize it either here

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreResult.cs:
```diff
@@ -171,7 +179,9 @@ public ISet<LibraryRange> GetAllUnresolved()
                 else
                 {
-                    log.LogMinimal($"Writing lock file to disk. Path: {result.LockFilePath}");
+                    log.LogMinimal(string.Format(CultureInfo.CurrentCulture,
```
I don't think any callers need this on Minimal anymore (there might be a test that verifies it is logged there). I suggest moving this to Information. Restore writes way too much into Minimal already

Same for tools

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs:
```diff
@@ -361,35 +361,38 @@ public static XDocument ReadExisting(string path, ILogger log)
-            string assetsFilePath,
-            bool restoreSuccess,
-            ILogger log)
+        public static string GetMSbuildFilePath(PackageSpec project, RestoreRequest request, string extension)
```
nit: MSB instead of MSb

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs:
```diff
@@ -361,35 +361,38 @@ public static XDocument ReadExisting(string path, ILogger log)
-
-                targetsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.targets");
-                propsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.props");
+                path = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.{extension}");
```
👍

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace NuGet.Commands
+{
+    class NoOpRestoreUtilities
```
use a modifier here, internal or public

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs:
```diff
@@ -0,0 +1,88 @@
+        internal static string GetToolCacheFilePath(RestoreRequest request, LockFile lockFile)
+        {
+
+            if(request.ProjectStyle != ProjectStyle.DotnetCliTool) { 
```
nit: format the document

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs:
```diff
@@ -0,0 +1,88 @@
+        /// </summary>
+        /// <param name="lockFile"></param>
+        /// <param name="request"></param>
+        /// <returns></returns>
```
The convention so far has been to delete empty tags from the summary. You don't have to fill them all, but if you do leave one empty then remove it completely.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs:
```diff
@@ -0,0 +1,88 @@
+
+            if (string.IsNullOrEmpty(projectCacheFilePath))
+            {
+                if (request.ProjectStyle == ProjectStyle.PackageReference
```
this looks like it can be simplified to *if tool -> get tool path  else -> get path* unless I'm missing something

emgarten commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -511,4 +511,40 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
+    <value>The restore inputs for '{0}' have not changed. No further actions are required to complete the restore.</value>
+  </data>
+  <data name="Log_SkippingLockFile" xml:space="preserve">
+    <value>Lock file has not changed. Skipping lock file write. Path: {0}</value>
```
Would you update this to say ``assets file`` instead of lock file? ``assets`` will be correct for 95% of the scenarios now, where as lock will be correct for 5%.  I think it's okay to have a single message for both.

emgarten commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -511,4 +511,40 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
+    <value>Lock file has not changed. Skipping lock file write. Path: {0}</value>
+  </data>
+  <data name="Log_ToolSkippingLockFile" xml:space="preserve">
+    <value>Tool lock file has not changed. Skipping lock file write. Path: {0}</value>
```
Tools are *always* ``assets files`` now.

shishirx34 committed:
Keep Project.json in sync with TPMinV for UWP Projects (#1341)

* sync project.json with framework from csproj

* fix frameworks

* cleanup

* Add tests

* add more test

* fix a bug

* address feedback

* always update framework if initially set for UAP

nkolev92 committed:
Address feedback + fix npe bug

nkolev92 commented:
@emgarten :bell:

nkolev92 committed:
Add the basic cache file tests

nkolev92 committed:
add missing file

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/CacheFile.cs:
```diff
@@ -38,7 +38,7 @@ public bool Equals(CacheFile other)
             }
 
-            return Version == other.Version && StringComparer.Ordinal.Equals(DgSpecHash, other.DgSpecHash);
+            return Version == other.Version && Success == other.Success &&StringComparer.Ordinal.Equals(DgSpecHash, other.DgSpecHash);
```
nit: spacing

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:
```diff
@@ -30,21 +30,21 @@ class NoOpRestoreResult : RestoreResult
+                        Strings.Log_ToolSkippingAssetsFile,
                         LockFilePath));
-                log.LogDebug(string.Format(CultureInfo.CurrentCulture,
+                log.LogInformation(string.Format(CultureInfo.CurrentCulture,
```
nit: I think the skip for tool and project could be Verbose. I don't think it needs to be under info, nuget.exe still prints at the info level and this might be a bit much to show all these.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreResult.cs:
```diff
@@ -59,11 +59,11 @@ public class RestoreResult : IRestoreResult
         /// Cache File path. The file path where the cache is written out
         /// </summary>
-        public string CacheFilePath { get;  }
+        protected string CacheFilePath { get;  }
```
👍

nkolev92 committed:
Don't include install.ps1 for VS14 (#1347)

nkolev92 committed:
Merge branch 'dev' into dev-nkolev92-noopcli

nkolev92 committed:
Address feedback

nkolev92 committed:
Fix commit msg tests

nkolev92 committed:
Add no op restore result tests

nkolev92 committed:
Fix last few tests
